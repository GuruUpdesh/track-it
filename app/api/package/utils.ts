import { NextResponse } from "next/server"

export function createErrorResponse(
	status: number,
	message: string
): NextResponse {
	return new NextResponse(
		JSON.stringify({
			error: message,
		}),
		{ status }
	)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function createSuccessResponse(data: any): NextResponse {
	return new NextResponse(JSON.stringify(data, null, 2), { status: 200 })
}
