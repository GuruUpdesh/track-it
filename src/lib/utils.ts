import { clsx, type ClassValue } from "clsx"
import { twMerge } from "tailwind-merge"

// from https://github.com/sadmann7/skateshop/blob/main/src/lib/utils.ts
export function cn(...inputs: ClassValue[]) {
	return twMerge(clsx(inputs))
}
