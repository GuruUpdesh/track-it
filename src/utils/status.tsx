import { AiOutlineWarning } from "react-icons/ai"
import { BsHouseDoor, BsQuestion, BsTruck } from "react-icons/bs"
import { TbTruckLoading } from "react-icons/tb"

type TStatusCode =
	| "PRE_TRANSIT"
	| "TRANSIT"
	| "DELIVERED"
	| "RETURNED"
	| "FAILURE"
	| "UNKNOWN"

interface StatusConfig {
	name:
		| "Pre Transit"
		| "Transit"
		| "Delivered"
		| "Returned"
		| "Failure"
		| "Unknown"
	code: TStatusCode
	icon: JSX.Element
	color: "red" | "yellow" | "lime" | "emerald" | "indigo"
}

interface StatusConfigs {
	[key: string]: StatusConfig
}

export const statusCodes = [
	"PRE_TRANSIT",
	"TRANSIT",
	"DELIVERED",
	"RETURNED",
	"FAILURE",
	"UNKNOWN",
]

const statusConfig: StatusConfigs = {
	pre_transit: {
		name: "Pre Transit",
		code: "PRE_TRANSIT",
		icon: <TbTruckLoading />,
		color: "yellow",
	},
	transit: {
		name: "Transit",
		code: "TRANSIT",
		icon: <BsTruck />,
		color: "lime",
	},
	delivered: {
		name: "Delivered",
		code: "DELIVERED",
		icon: <BsHouseDoor />,
		color: "emerald",
	},
	returned: {
		name: "Returned",
		code: "RETURNED",
		icon: <BsTruck style={{ transform: "scaleX(-1)" }} />,
		color: "emerald",
	},
	failure: {
		name: "Failure",
		code: "FAILURE",
		icon: <AiOutlineWarning />,
		color: "red",
	},
	unknown: {
		name: "Unknown",
		code: "UNKNOWN",
		icon: <BsQuestion />,
		color: "indigo",
	},
}

export const statuses = Object.keys(statusConfig).map((status) =>
	getStatus(status as TStatusCode)
)

export function getStatus(statusCode: TStatusCode) {
	return statusConfig[statusCode.toLowerCase()]
}
