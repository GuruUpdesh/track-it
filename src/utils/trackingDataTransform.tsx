import {
	TTrackingData,
	TShippoResponse,
	TShippoTrackingHistory,
	TStatus,
	THistoryRow,
	TRawLocation,
} from "@/app/api/package/typesAndSchemas"
import {
	endOfDay,
	isEqual,
	startOfDay,
	format,
	differenceInCalendarDays,
	formatDistance,
} from "date-fns"
import {
	convertLocationObjectToString,
	estimateProgress,
	extractDeliveryLocation,
} from "./package"
import { getLatLng } from "@/app/api/track/route"

export function simplifyDetailMessage(
	message: string,
	status: TStatus
): string {
	const lowercaseMessage = message.toLowerCase()

	if (status === "DELIVERED") {
		return "Delivered"
	}

	if (status === "TRANSIT") {
		if (lowercaseMessage.includes("departed")) {
			return "Departed"
		} else if (lowercaseMessage.includes("arrived")) {
			return "Arrived at facility"
		} else if (lowercaseMessage.includes("in transit")) {
			return "In transit"
		} else if (lowercaseMessage.includes("out for delivery")) {
			return "Out for delivery"
		}
	}

	if (status === "PRE_TRANSIT") {
		return "Shipment information received"
	}

	return message
}

export function getEta(eta: string | null): string | null {
	if (!eta) {
		return null
	}

	const etaDate = new Date(eta)

	if (isEqual(etaDate, startOfDay(etaDate))) {
		const endOfEtaDay = endOfDay(etaDate)
		return format(endOfEtaDay, "yyyy-MM-dd HH:mm:ss")
	} else {
		return eta
	}
}

export function getSourceAndDestinationLocations(packageInfo: TShippoResponse) {
	let source = packageInfo.address_from
	const destination = packageInfo.address_to

	const trackingHistory = packageInfo.tracking_history
	if (trackingHistory.length === 0) {
		source = null
	} else {
		for (let i = 0; i < trackingHistory.length; i++) {
			const historyItem = trackingHistory[i]
			if (historyItem.location) {
				source = historyItem.location
				break
			}
		}
	}

	if (!source && !destination) {
		return null
	}

	const sourceString = convertLocationObjectToString(source)
	const destinationString = convertLocationObjectToString(destination)

	if (!source) {
		return `Coming to ${destinationString}`
	} else if (!destination) {
		return `Coming from ${sourceString}`
	} else {
		if (sourceString === destinationString) {
			return `Coming to ${destinationString}`
		}
		return `Coming from ${sourceString} to ${destinationString}`
	}
}

export function getTransitTime(packageInfo: TShippoResponse) {
	const trackingHistory = packageInfo.tracking_history
	if (trackingHistory.length < 1) {
		return null
	}
	const startDate = new Date(trackingHistory[0].status_date)
	let endDate = new Date()
	if (trackingHistory.length > 1) {
		endDate = new Date(
			trackingHistory[trackingHistory.length - 1].status_date
		)
	}

	if (Math.abs(differenceInCalendarDays(endDate, startDate)) < 1) {
		return null
	}

	const transitTime = formatDistance(startDate, endDate)

	const trackingStatus = packageInfo.tracking_status.status
	if (trackingStatus === "DELIVERED" || trackingStatus === "RETURNED") {
		return "Your package was in transit for " + transitTime
	}
	return "Your package has been in transit for " + transitTime
}

export async function simplifyRawTrackingData(
	data: TShippoResponse
): Promise<TTrackingData> {
	async function simplifyTrackingHistory(
		trackingHistory: TShippoTrackingHistory
	): Promise<THistoryRow> {
		const locationName = convertLocationObjectToString(
			trackingHistory.location
		)
		let lat: number | null = null
		let lng: number | null = null
		try {
			const latLng = await getLatLng(trackingHistory.location)
			lat = latLng.lat
			lng = latLng.lng
		} catch (error) {
			console.error(error)
		}
		return {
			status: trackingHistory.status,
			detailedStatus: trackingHistory.status_details,
			location: {
				name: locationName,
				lat,
				lng,
			},
			date: trackingHistory.status_date,
			deliveryLocation: extractDeliveryLocation(
				trackingHistory.status_details
			),
		}
	}

	const start = new Date().getTime()
	const simplifiedHistoryRows = await Promise.all(
		data.tracking_history.map(simplifyTrackingHistory)
	)

	const simplifiedTrackingHistory: TTrackingData = {
		trackingNumber: data.tracking_number,
		courier: data.carrier,
		eta: getEta(data.eta),
		progressPercentage: 0,
		sourceAndDestinationString: getSourceAndDestinationLocations(data),
		transitTime: getTransitTime(data),
		status: await simplifyTrackingHistory(
			data.tracking_status as TShippoTrackingHistory
		),
		service: data.servicelevel.name,
		trackingHistory: simplifiedHistoryRows,
	}

	simplifiedTrackingHistory.progressPercentage = estimateProgress(
		simplifiedTrackingHistory.eta,
		simplifiedTrackingHistory.status.status,
		simplifiedTrackingHistory.trackingHistory[0].date
	)

	const end = new Date().getTime()
	console.log(`Google API: ${end - start}ms`)

	return simplifiedTrackingHistory
}

export function formatRawLocation(location: TRawLocation): string {
	if (!location) {
		return ""
	}

	const { city, state, zip, country } = location
	const parts = []

	if (city) {
		parts.push(city)
	}

	if (state) {
		parts.push(state)
	}

	if (zip) {
		parts.push(zip)
	}

	if (country) {
		parts.push(country)
	}

	return parts.join(", ")
}
