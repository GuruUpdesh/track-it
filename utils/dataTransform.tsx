import {
	ShippoResponse,
	ShippoTrackingHistory,
	TStatus,
	TrackingHistory,
} from "@/app/api/package/typesAndSchemas"
import {
	endOfDay,
	isEqual,
	startOfDay,
	format,
	differenceInCalendarDays,
	formatDistance,
} from "date-fns"
import {
	convertLocationObjectToString,
	extractDeliveryLocation,
} from "./package"

export function simplifyDetailMessage(
	message: string,
	status: TStatus
): string {
	const lowercaseMessage = message.toLowerCase()

	if (status === "DELIVERED") {
		return "Delivered"
	}

	if (status === "TRANSIT") {
		if (lowercaseMessage.includes("departed")) {
			return "Departed"
		} else if (lowercaseMessage.includes("arrived")) {
			return "Arrived at facility"
		} else if (lowercaseMessage.includes("in transit")) {
			return "In transit"
		} else if (lowercaseMessage.includes("out for delivery")) {
			return "Out for delivery"
		}
	}

	if (status === "PRE_TRANSIT") {
		return "Shipment information received"
	}

	return message
}

export function getEta(eta: string | null): string | null {
	if (!eta) {
		return null
	}

	const etaDate = new Date(eta)

	if (isEqual(etaDate, startOfDay(etaDate))) {
		const endOfEtaDay = endOfDay(etaDate)
		return format(endOfEtaDay, "yyyy-MM-dd HH:mm:ss")
	} else {
		return eta
	}
}

export function getSourceAndDestinationLocations(packageInfo: ShippoResponse) {
	let source = packageInfo.address_from
	let destination = packageInfo.address_to

	const trackingHistory = packageInfo.tracking_history
	if (trackingHistory.length === 0) {
		source = null
	} else {
		for (let i = 0; i < trackingHistory.length; i++) {
			const historyItem = trackingHistory[i]
			if (historyItem.location) {
				source = historyItem.location
				break
			}
		}
	}

	if (!source && !destination) {
		return null
	}

	const sourceString = convertLocationObjectToString(source)
	const destinationString = convertLocationObjectToString(destination)

	if (!source) {
		return `Coming to ${destinationString}`
	} else if (!destination) {
		return `Coming from ${sourceString}`
	} else {
		if (sourceString === destinationString) {
			return `Coming to ${destinationString}`
		}
		return `Coming from ${sourceString} to ${destinationString}`
	}
}

export function getTransitTime(packageInfo: ShippoResponse) {
	const trackingHistory = packageInfo.tracking_history
	if (trackingHistory.length < 1) {
		return null
	}
	const startDate = new Date(trackingHistory[0].status_date)
	let endDate = new Date()
	if (trackingHistory.length > 1) {
		endDate = new Date(
			trackingHistory[trackingHistory.length - 1].status_date
		)
	}

	if (Math.abs(differenceInCalendarDays(endDate, startDate)) < 1) {
		return null
	}

	const transitTime = formatDistance(startDate, endDate)

	const trackingStatus = packageInfo.tracking_status.status
	if (trackingStatus === "DELIVERED" || trackingStatus === "RETURNED") {
		return "Your package was in transit for " + transitTime
	}
	return "Your package has been in transit for " + transitTime
}

export function simplifyTrackingHistory(
	trackingHistory: ShippoTrackingHistory
): TrackingHistory {
	return {
		status: trackingHistory.status,
		detailedStatus: simplifyDetailMessage(
			trackingHistory.status_details,
			trackingHistory.status
		),
		location: convertLocationObjectToString(trackingHistory.location),
		date: trackingHistory.status_date,
		deliveryLocation: extractDeliveryLocation(
			trackingHistory.status_details
		),
	}
}
