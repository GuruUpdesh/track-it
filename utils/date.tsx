import { format, formatDistance } from "date-fns"

export function formatRelativeDate(date: string): string {
	const dateObject = new Date(date)
	const relative = formatDistance(dateObject, new Date(), { addSuffix: true })
	return relative
}

export function formatDate(date: string): string {
	const dateObject = new Date(date)
	return format(dateObject, "MMMM d, yyyy") // e.g., January 1, 2023
}

export function getTimeFromDate(date: string): string {
	const dateObject = new Date(date)
	return format(dateObject, "h:mm aa") // e.g., 2:30 PM
}
