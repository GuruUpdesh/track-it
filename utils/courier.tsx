import { TCourier } from "@/app/api/package/typesAndSchemas"
import { FaFedex, FaUps } from "react-icons/fa"
import { MdOutlineExplore } from "react-icons/md"
import { SiUsps } from "react-icons/si"

import type { JSX } from "react";

const UPS_REGEX_PATTERN =
	/\b(1Z ?[0-9A-Z]{3} ?[0-9A-Z]{3} ?[0-9A-Z]{2} ?[0-9A-Z]{4} ?[0-9A-Z]{3} ?[0-9A-Z]|T\d{3} ?\d{4} ?\d{3})\b/i

const USPS_REGEX_PATTERN_1 =
	/\b((420 ?\d{5} ?)?(91|92|93|94|95|01|03|04|70|23|13)\d{2} ?\d{4} ?\d{4} ?\d{4} ?\d{4}( ?\d{2,6})?)\b/i
const USPS_REGEX_PATTERN_2 =
	/\b((M|P[A-Z]?|D[C-Z]|LK|E[A-C]|V[A-Z]|R[A-Z]|CP|CJ|LC|LJ) ?\d{3} ?\d{3} ?\d{3} ?[A-Z]?[A-Z]?)\b/i
const USPS_REGEX_PATTERN_3 = /\b(82 ?\d{3} ?\d{3} ?\d{2})\b/i

const ONTRAC_REGEX_PATTERN = /\b(C\d{14})\b/i

const FEDEX_REGEX_PATTERN =
	/\b(((96\d\d|6\d)\d{3} ?\d{4}|96\d{2}|\d{4}) ?\d{4} ?\d{4}( ?\d{3}|\d{15})?)\b/i

interface CourierConfig {
	name: string
	code: TCourier
	icon: JSX.Element
	tracking_url: string
	patterns: RegExp[]
}

interface CourierConfigs {
	[key: string]: CourierConfig
}

export const courierCodes: TCourier[] = [
	"ups",
	"usps",
	"ontrac",
	"fedex",
	"shippo",
]

const courierConfig: CourierConfigs = {
	ups: {
		name: "UPS",
		code: "ups",
		icon: <FaUps />,
		tracking_url: "https://www.ups.com/mobile/track?trackingNumber=",
		patterns: [new RegExp(UPS_REGEX_PATTERN)],
	},
	usps: {
		name: "USPS",
		code: "usps",
		icon: <SiUsps />,
		tracking_url:
			"https://tools.usps.com/go/TrackConfirmAction?qtc_tLabels1=",
		patterns: [
			new RegExp(USPS_REGEX_PATTERN_1),
			new RegExp(USPS_REGEX_PATTERN_2),
			new RegExp(USPS_REGEX_PATTERN_3),
		],
	},
	ontrac: {
		name: "OnTrac",
		code: "ontrac",
		icon: <MdOutlineExplore />,
		tracking_url: "http://www.ontrac.com/trackres.asp?tracking_number=",
		patterns: [new RegExp(ONTRAC_REGEX_PATTERN)],
	},
	fedex: {
		name: "FedEx",
		code: "fedex",
		icon: <FaFedex />,
		tracking_url: "https://www.fedex.com/fedextrack/?tracknumbers=",
		patterns: [new RegExp(FEDEX_REGEX_PATTERN)],
	},
	shippo: {
		name: "Shippo",
		code: "shippo",
		icon: <MdOutlineExplore />,
		tracking_url: "https://goshippo.com/tracking/",
		patterns: [],
	},
}

export const couriers = Object.keys(courierConfig).map((courier) =>
	getCourier(courier as TCourier)
)

export function getCourier(courierCode: TCourier) {
	return courierConfig[courierCode]
}

export function getCourierFromTrackingNumber(trackingNumber: string) {
	const matchingCouriers: TCourier[] = []
	for (const courier in courierConfig) {
		for (const pattern of courierConfig[courier].patterns) {
			if (pattern.test(trackingNumber)) {
				matchingCouriers.push(courier as TCourier)
			}
		}
	}

	if (matchingCouriers.length === 1) {
		return getCourier(matchingCouriers[0])
	} else if (matchingCouriers.length > 1) {
		console.warn(
			"Multiple couriers found for tracking number, returning first"
		)
		return getCourier(matchingCouriers[0])
	}

	return null
}
