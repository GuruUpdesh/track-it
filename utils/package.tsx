import { TLocation, TStatus } from "@/app/api/package/typesAndSchemas"
import { format, formatDistance } from "date-fns"
import { AiOutlineWarning } from "react-icons/ai"
import { BsHouseDoor, BsMailbox, BsQuestion, BsTruck } from "react-icons/bs"
import { TbTruckLoading } from "react-icons/tb"

export function formatRelativeDate(date: string): string {
	const dateObject = new Date(date)
	const relative = formatDistance(dateObject, new Date(), { addSuffix: true })
	return relative
}

export function formatDate(date: string): string {
	const dateObject = new Date(date)
	return format(dateObject, "MMMM d, yyyy") // e.g., January 1, 2023
}

export function getTimeFromDate(date: string): string {
	const dateObject = new Date(date)
	return format(dateObject, "h:mm aa") // e.g., 2:30 PM
}

export function convertLocationObjectToString(
	location: TLocation | null
): string {
	if (!location) return "Location not found"

	let locationString = ""
	if (location.city) locationString += location.city
	if (location.state) {
		if (!location.city) locationString += location.state
		else locationString += `, ${location.state}`
	}
	if (location.country && !location.state) {
		if (!location.city) locationString += location.country
		else locationString += `, ${location.country}`
	}

	return locationString
}

export function extractDeliveryLocation(statusUpdate: string): string | null {
	const deliveredRegex = /delivered/i
	const frontDoorRegex = /(front door|doorstep|entrance|gate)/i
	const mailboxRegex = /mailbox/i

	if (deliveredRegex.test(statusUpdate)) {
		if (frontDoorRegex.test(statusUpdate)) {
			return "home"
		} else if (mailboxRegex.test(statusUpdate)) {
			return "mailbox"
		}
	}

	return null
}

export function getIconForStatus(
	status: TStatus,
	deliveryLocation?: string | null
): JSX.Element {
	switch (status) {
		case "PRE_TRANSIT":
			return <TbTruckLoading />
		case "TRANSIT":
			return <BsTruck />
		case "DELIVERED":
			if (deliveryLocation === "mailbox") return <BsMailbox />
			return <BsHouseDoor />
		case "RETURNED":
			return <BsTruck style={{ transform: "scaleX(-1)" }} />
		case "FAILURE":
			return <AiOutlineWarning />
		case "UNKNOWN":
			return <BsQuestion />
		default:
			return <></>
	}
}
