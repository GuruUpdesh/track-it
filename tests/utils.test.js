import {
	getCourierStringFromCode,
	getCouriersFromTrackingNumber,
} from "../utils/courier"
import { formatDate, getTimeFromDate } from "../utils/date"
import {
	convertLocationObjectToString,
	extractDeliveryLocation,
} from "../utils/package"

describe("date utils", () => {
	test("formatDate returns formatted date (same year)", () => {
		const date = new Date(2023, 0, 1).toISOString() // January 1, 2023
		expect(formatDate(date)).toBe("Jan 1")
	})

	test("formatDate returns formatted date (different year)", () => {
		const date = new Date(2022, 0, 1).toISOString() // January 1, 2023
		expect(formatDate(date)).toBe("Jan 1, 2022")
	})

	test("getTimeFromDate returns formatted time", () => {
		const date = new Date(2023, 0, 1, 14, 30).toISOString() // January 1, 2023, 14:30
		expect(getTimeFromDate(date)).toBe("2:30pm")
	})
})

describe("package utils", () => {
	describe("Testing location conversion function", () => {
		test("convertLocationObjectToString returns formatted location", () => {
			const locations = [
				{
					input: { city: "New York", state: "NY" },
					output: "New York, NY",
				},
				{ input: { state: "NY" }, output: "NY" },
				{ input: { city: "New York" }, output: "New York" },
				{ input: { country: "US" }, output: "US" },
				{
					input: { country: "US", city: "New York" },
					output: "New York, US",
				},
				{
					input: { city: "New York", state: "NY", country: "US" },
					output: "New York, NY",
				},
			]
			for (const location of locations) {
				expect(convertLocationObjectToString(location.input)).toBe(
					location.output
				)
			}
		})

		test("convertLocationObjectToString returns error string for null", () => {
			expect(convertLocationObjectToString(null)).toBe(
				"Location not found"
			)
		})
	})

	describe("Testing delivery location extraction function", () => {
		const testCases = [
			{
				input: "The package was delivered at the front door",
				expected: "home",
			},
			{
				input: "The package was delivered in the mailbox",
				expected: "mailbox",
			},
			{ input: "The package was delivered", expected: null },
			{
				input: "The package was delivered at the Front Door",
				expected: "home",
			},
			{
				input: "The package was delivered   at the front door   ",
				expected: "home",
			},
			{
				input: "The package was delivered at the front door #123",
				expected: "home",
			},
			{
				input: "The package was delivered at the front door - Apt 123",
				expected: "home",
			},
			{
				input: "The package was delivered at the front door, Apt 123",
				expected: "home",
			},
			{
				input: "The package was delivered at the front door; Apt 123",
				expected: "home",
			},
			{
				input: "The package was delivered at the front door: Apt 123",
				expected: "home",
			},
			{
				input: "The package was delivered at the doorstep",
				expected: "home",
			},
			{
				input: "The package was delivered at the entrance",
				expected: "home",
			},
			{
				input: "The package was delivered at the gate",
				expected: "home",
			},
		]

		testCases.forEach(({ input, expected }) => {
			test(`extractDeliveryLocation identifies delivery location: ${input}`, () => {
				expect(extractDeliveryLocation(input)).toBe(expected)
			})
		})
	})
})

describe("courier utils", () => {
	describe("Testing getCourierStringFromCode function", () => {
		test("Valid courier code returns correct string", () => {
			expect(getCourierStringFromCode("ups")).toBe("UPS")
			expect(getCourierStringFromCode("usps")).toBe("USPS")
			expect(getCourierStringFromCode("ontrac")).toBe("OnTrac")
			expect(getCourierStringFromCode("fedex")).toBe("FedEx")
			expect(getCourierStringFromCode("shippo")).toBe("test")
		})

		test("Invalid courier code returns 'Invalid Courier'", () => {
			console.error = jest.fn() // Mocking console.error to prevent error logging
			expect(getCourierStringFromCode("invalid")).toBe("Invalid Courier")
			expect(console.error).toHaveBeenCalledWith(
				"Invalid Courier Code",
				"invalid"
			)
		})
	})
	describe("Testing getCouriersFromTrackingNumber function", () => {
		test("Valid tracking number returns matching couriers", () => {
			const testCases = [
				{
					trackingNumber: "1Z1234567890123456",
					expectedCouriers: ["ups"],
				},
				{
					trackingNumber: "T1234567890",
					expectedCouriers: ["ups"],
				},
				{
					trackingNumber: "9112345678901234567890",
					expectedCouriers: ["usps"],
				},
				{
					trackingNumber: "9374889696021505850386",
					expectedCouriers: ["usps"],
				},
				{
					trackingNumber: "C12345678901234",
					expectedCouriers: ["ontrac"],
				},
				{
					trackingNumber: "9601234567890123456789",
					expectedCouriers: ["fedex"],
				},
				{
					trackingNumber: "9400110200793594709945",
					expectedCouriers: ["usps"],
				},
				{
					trackingNumber: "9205510200793594709946",
					expectedCouriers: ["usps"],
				},
				{
					trackingNumber: "9407100200793594709947",
					expectedCouriers: ["usps"],
				},
				{
					trackingNumber: "999999999999",
					expectedCouriers: ["fedex"],
				},
				{
					trackingNumber: "999999999999999",
					expectedCouriers: ["fedex"],
				},
			]

			for (const testCase of testCases) {
				const { trackingNumber, expectedCouriers } = testCase
				const couriers = getCouriersFromTrackingNumber(trackingNumber)
				expect(couriers).toEqual(
					expect.arrayContaining(expectedCouriers)
				)
			}
		})

		test("Invalid tracking number returns an empty array", () => {
			const trackingNumber = "invalid"
			const matchingCouriers =
				getCouriersFromTrackingNumber(trackingNumber)
			expect(matchingCouriers).toEqual([])
		})
	})
})
