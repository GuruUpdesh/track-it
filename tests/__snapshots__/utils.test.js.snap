// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing status icon function getIconForStatus returns correct icons 1`] = `
<div>
  <svg
    fill="none"
    height="1em"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <path
      d="M2 3h1a2 2 0 0 1 2 2v10a2 2 0 0 0 2 2h15"
    />
    <path
      d="M9 6m0 3a3 3 0 0 1 3 -3h4a3 3 0 0 1 3 3v2a3 3 0 0 1 -3 3h-4a3 3 0 0 1 -3 -3z"
    />
    <path
      d="M9 19m-2 0a2 2 0 1 0 4 0a2 2 0 1 0 -4 0"
    />
    <path
      d="M18 19m-2 0a2 2 0 1 0 4 0a2 2 0 1 0 -4 0"
    />
  </svg>
</div>
`;
